Nos/Service/PushNotificationRegistrar.swift

```diff
@@ -0,0 +1,108 @@
+        case missingPubkey
+        case missingDeviceToken
+        case unexpected
+        
+        var errorDescription: String? {
+            switch self {
+            case .missingPubkey:        "The pubkey is required"
+            case .missingDeviceToken:   "Device token is required"
+            case .unexpected:           "Something unexpected happened"
+            }
+        }
+    }
+    
+#if DEBUG
+    private static let notificationServiceRelayURL = URL(string: "wss://dev-notifications.nos.social")!
+#else
+    private static let notificationServiceRelayURL = URL(string: "wss://notifications.nos.social")!
+#endif
+    
+    @Dependency(\.relayService) private var relayService
+    @Dependency(\.analytics) private var analytics
+    @Dependency(\.currentUser) private var currentUser
+    
+    private var deviceToken: Data?
+    private var registeredPubkey: String?
+    private var registrationTask: Task<Void, Error>?
+    
+    /// Registers a user for push notifications by publishing a registration event. Calls to
+    /// this function are handled serially.
+    /// - Parameters:
+    ///   - user: The user to register for notifications.
+    ///   - deviceToken: The device token to register with. A cached token will be used if none is passed in.
+    ///   - context: The Core Data context to use.
+    func register(_ user: CurrentUser, with deviceToken: Data? = nil, context: NSManagedObjectContext) async throws {
+        guard let userKey = user.publicKeyHex, let keyPair = user.keyPair else {
+            throw PushNotificationError.missingPubkey
+        }
+        
+        try await registrationTask?.value   // wait for the previous task to finish
+        
+        guard let token = deviceToken ?? self.deviceToken else {
+            throw PushNotificationError.missingDeviceToken
+        }
+        
+        self.deviceToken = token
+        
+        guard userKey != registeredPubkey else {
+            return  // already registered this user
+        }
+        
+        registrationTask = Task {
+            do {
+                let jsonEvent = JSONEvent(
+                    pubKey: userKey,
+                    kind: .notificationServiceRegistration,
+                    tags: [],
+                    content: try await createRegistrationContent(deviceToken: token, user: user)
+                )
+                try await relayService.publish(
+                    event: jsonEvent,
+                    to: Self.notificationServiceRelayURL,
+                    signingKey: keyPair,
+                    context: context
+                )
+                registeredPubkey = userKey
+            } catch {
+                analytics.pushNotificationRegistrationFailed(reason: error.localizedDescription)
+                throw error
+            }
+        }
+        
+        try await registrationTask?.value
+    }
+    
+    /// Builds the string needed for the `content` field in the special
```

Comment:
Is this doc comment missing something? I'm not quite understanding what it means.

Reply:
This was simply copied over from the original. I'll just remove the "in the special" part since it doesn't make sense.

Reply:
Ah yes, sorry for making you deal with that!